WalletWasabi/Wallets/WalletManager.cs

```diff
@@ -85,47 +100,118 @@ public bool AnyWallet()
 			}
 		}
 
-		public async Task<Wallet> CreateAndStartWalletAsync(KeyManager keyManager)
+		public async Task<Wallet> StartWalletAsync(Wallet wallet)
 		{
-			using (await AddRemoveLock.LockAsync(CancelAllInitialization.Token).ConfigureAwait(false))
+			Guard.NotNull(nameof(wallet), wallet);
+
+			if (wallet.State == WalletState.Starting)
+			{
+				return wallet;
+			}
+
+			using (await StartStopWalletLock.LockAsync(CancelAllInitialization.Token).ConfigureAwait(false))
 			{
-				Wallet wallet = null;
 				try
 				{
-					wallet = new Wallet(Network, BitcoinStore, keyManager, Synchronizer, Nodes, WalletDirectories.WorkDir, ServiceConfiguration, FeeProvider, BitcoinCoreNode);
+					wallet.RegisterServices(BitcoinStore, Synchronizer, Nodes, ServiceConfiguration, FeeProvider, BitcoinCoreNode);
 
 					var cancel = CancelAllInitialization.Token;
-					lock (Lock)
-					{
-						Wallets.Add(wallet, new HashSet<uint256>());
-					}
-
 					Logger.LogInfo($"Starting {nameof(Wallet)}...");
 					await wallet.StartAsync(cancel).ConfigureAwait(false);
 					Logger.LogInfo($"{nameof(Wallet)} started.");
-
 					cancel.ThrowIfCancellationRequested();
 
-					wallet.WalletRelevantTransactionProcessed += TransactionProcessor_WalletRelevantTransactionProcessed;
-					wallet.OnDequeue += ChaumianClient_OnDequeue;
-
 					return wallet;
 				}
 				catch
 				{
-					if (wallet is { })
+					await wallet.StopAsync(CancellationToken.None).ConfigureAwait(false);
+					throw;
+				}
+			}
+		}
+
+		public Task<Wallet> StartWalletAsync(KeyManager keyManagerToFindByReference)
+		{
+			Wallet wallet;
+			lock (Lock)
+			{
+				wallet = Wallets.Single(x => x.Key.KeyManager == keyManagerToFindByReference).Key;
+			}
+
+			return StartWalletAsync(wallet);
+		}
+
+		public Task<Wallet> AddAndStartWalletAsync(KeyManager keyManager)
+		{
+			var wallet = AddWallet(keyManager);
+			return StartWalletAsync(wallet);
+		}
+
+		public Wallet AddWallet(KeyManager keyManager)
+		{
+			lock (Lock)
+			{
+				if (Wallets.Any(w => w.Key.KeyManager == keyManager))
```

Comment:
Why is the `Wallets` defined as a `Dictionary<Wallet, HashSet<uint256>>`? I can see where the HashSet is used.

Reply:
It is used here: https://github.com/molnard/WalletWasabi/blob/c74ed79fa08e2e5de0d96e9f23f5c5ab5fd3ef26/WalletWasabi/Wallets/WalletManager.cs#L299

I am sure we can remove this in the future but it would be too much change by now.